### PR TITLE
Migration of content away from existing SubServicePages

### PR DESCRIPTION
**Pull request by engineervix**
[Link to Ticket](https://torchbox.monday.com/boards/1192293412/pulses/1242063350)

### Description of Changes Made

**TLDR;**

This PR builds up on work done in #516 by adding some tooling to help us migrate content away from SubServicePage models to SubPropositionPage models.

**Rationale**

The idea here is to automate the process of migrating content as much as possible, while leaving room for manual checking and cleanup afterwards (which should be minimal, as all the heavylifting should already have been done).

> **Note**
>
> Running this on production should ideally be done when the site is not very busy — it might be safer to temporarily put the site in maintenance mode.

**Details**

There are two management commands:

1. `./manage.py migrate_subservicepages` → running this does the following:
    - check if a migration has already been run. If it has, do nothing, otherwise, proceed.
    - for each SubServicePage:
      - copy all the content so that we can use it to create a new SubPropositionPage
      - add a suffix to the title and slug, then unpublish the page if it was live
      - create a corresponding SubPropositionPage at the same location, with all the content copied earlier
    - create a record of the migration in the database, which is used to determine whether a migration has already been run.
    - check all content for references to the SubServicePages, and replace these references with the new SubPropositionPages. This is done in
      - StreamFields
      - RichTextFields
      - ~~ForeignKeys to `"wagtailcore.Page"` or `"services.SubServicePage"`~~ not done due to time constraints
      - ~~Settings models (Site settings, Navigation settings)~~ not necessary as SubServicePages aren't referenced here
    - optionally send CSV reports via email to any specified recipients. This reports documents changes made, and areas which need manual intervention.
2. `./manage.py undo_subservicepage_migration` → this reverses the changes made in 1. above.

> **Warning**
>
> 1. Any draft changes on SubServicePages will not be transferred to the new SubPropositionPage. This will have to be done manually.
> 2. All revisions will not be transferred to the new page.
> 3. For the undo operation, we have no control on changes made to pages after the migration.

**Additional info**

> **Note**
>
> - I added a commit (372427f1) to replace `propositions.TestimonialBlock` with `core.TestimonialBlock`, because of changes made in 6518fb44, #516   
> - I added a commit (7f19575) to fix WorkPage `date` field. 

### How to Test

1. Create a backup of the database (locally, you could use [DSLR](https://github.com/mixxorz/DSLR))
2. If running locally, ensure you have staging data. Go to Wagtail Admin and take note of the SubServicePages
3. Then

    ````bash
    # omit `--send-to ...` if you don't wanna email a report
    # note that you can send report to multiple recipients by adding more
    # email addresses separated by a space 
    ./manage.py migrate_subservicepages --send-to name@example.com
    ```

4. Go to wagtail admin and check that 

    - all the SubServicePages are now unpublished, and have a suffix added to their titles and slugs
    - for each unpublished SubServicePage, there's a corresponding SubPropositionPage at the same location

5. View the new SubPropositionPages and compare them with the SubServicePages -- they should be identical
6. Take note of the pages which require manual checking of links to avert 404s (there should only be 2 pages with the following titles
    - "An Extra 3 Million Organic Clicks a Month for the NHS Website"
    - "Raising millions for Islamic Relief UK during Ramadan"
8. Need to find a way to check for 404s on the entire site

To undo migrations

```bash
manage.py undo_subservicepage_migration
```

For easier testing of emails locally, you could use [maildev](https://github.com/maildev/maildev)

Add the following to `docker-compose.yaml`

```yaml
  maildev:
    image: maildev/maildev
    ports:
      - "1080:1080"
```

And update `settings/local.py` as follows

```py
EMAIL_BACKEND = "django.core.mail.backends.smtp.EmailBackend"

# https://docs.djangoproject.com/en/4.2/ref/settings/#email-host
EMAIL_HOST = "maildev"
# https://docs.djangoproject.com/en/4.2/ref/settings/#email-port
EMAIL_PORT = 1025
```

### Screenshots

<details>
  <summary>./manage.py migrate_subservicepages</summary>

![image](https://github.com/torchbox/wagtail-torchbox/assets/7713776/96236f15-2588-42c9-a174-6a8cb57d9b66)

![image](https://github.com/torchbox/wagtail-torchbox/assets/7713776/755c0e60-7a0f-4f67-83b1-8005f3b8d3ee)

![image](https://github.com/torchbox/wagtail-torchbox/assets/7713776/e9dd79b3-bfd1-46c1-a2da-6ffa674d573d)

</details>

<details>
  <summary>./manage.py undo_subservicepage_migration</summary>

![image](https://github.com/torchbox/wagtail-torchbox/assets/7713776/da4352c7-ce1f-40be-a6aa-2ac3bd1bfa2e)

</details>

<details>
  <summary>Wagtail Admin</summary>

![image](https://github.com/torchbox/wagtail-torchbox/assets/7713776/f7d6f1ff-8753-4fa8-956d-ec526554bf83)

</details>

<details>
  <summary>Page comparison</summary>

![image](https://github.com/torchbox/wagtail-torchbox/assets/7713776/bb225f35-1653-4173-b659-627ab58c5e58)

![image](https://github.com/torchbox/wagtail-torchbox/assets/7713776/bee7f6b7-e5d0-4001-854b-076069e62fb6)

![image](https://github.com/torchbox/wagtail-torchbox/assets/7713776/6aad2590-9956-42a7-8b55-d2b929a39207)

![image](https://github.com/torchbox/wagtail-torchbox/assets/7713776/de425e3a-f805-46cd-9ba0-e27f9184edab)


</details>

### MR Checklist

- [x] Add a description of your pull request and instructions for the reviewer to verify your work.
- [x] If your pull request is for a specific ticket, link to it in the description.
- [x] Stay on point and keep it small so the merge request can be easily reviewed.
- [ ] Consider adding unit tests, especially for bug fixes. If you don't, tell us why.
- [x] Tests and linting passes.
- [ ] Consider updating documentation. If you don't, tell us why.
- [ ] If relevant, list the environments / browsers in which you tested your changes.

### CODE REVIEW

**Review by victoriachan — CHANGES_REQUESTED** [review]
Hi Victor,

Wow, what a lot of code here. Well done for tackling it.

When testing locally with staging data, I encountered an error to do with the old testimonial streamfield (`AttributeError: 'StructValue' object has no attribute 'quote'`). I think we need to be more pessimistic about the existing streamfield data. We may need more exception handling added.

When the error was encountered, the command just failed halfway. I reverted it, but the revert didn't work fully as the page with the error was never added to the migration table. 

Hope it won't be too much work to make these changes, but as this concerns database changes, we need to be more cautious.

File: tbx/propositions/constants.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# These suffixes are used in the
+# sub service to sub proposition page migration.
+# See:
+# - tbx/propositions/management/commands/migrate_subservicepages.py
+# - tbx/propositions/management/commands/undo_subservicepage_migration.py
+TITLE_SUFFIX = "(NO LONGER IN USE)"
```
[the remark this answers is not in the document]
Comment (engineervix) [reply tbx/propositions/constants.py]: Thanks, updated in 5db0bfba

File: tbx/propositions/management/commands/migrate_subservicepages.py [hunk 2]
```diff
@@ -0,0 +1,1005 @@
...
+                # the source & target blocks are slightly different
+                # so we need to add this to the the `testimonials` block's `testimonials` field
+                data = {
+                    "title": "Clients",  # this is the default title
+                    "testimonials": [
+                        {
+                            "quote": testimonial.quote,
+                            "name": testimonial.name,
+                            "role": testimonial.role,
+                            "link": testimonial.link if testimonial.link else None,
```
[the remark this answers is not in the document]
Comment (engineervix) [reply tbx/propositions/management/commands/migrate_subservicepages.py]: Nice catch! I thought I was using latest staging data but I actually wasn't so I didn't encounter this when testing. After pulling in latest data I encountered the same error, and have fixed it in `9bcb2cbd`

File: tbx/propositions/management/commands/migrate_subservicepages.py [hunk 3]
```diff
@@ -0,0 +1,1005 @@
...
+            else:
+                new_data[key] = update_stream_data(value)
+        return new_data
+
+    else:
+        return data
+
+
+def update_page_links():
+    pages = Page.objects.not_type(SubServicePage).specific()
```
[the remark this answers is not in the document]
Comment (engineervix) [reply tbx/propositions/management/commands/migrate_subservicepages.py:581]: Since `update_page_links()` is executed after the SubPropositionPage has already been created from a SubServicePage, the new SubPropositionPage will be updated if it contains a link to another SubServicePage

File: tbx/propositions/management/commands/migrate_subservicepages.py [hunk 4]
```diff
@@ -0,0 +1,1005 @@
...
+            first_published_at = old_page.first_published_at
+            last_published_at = old_page.last_published_at
+            locked = old_page.locked
+            # locked_* fields only apply to locked pages
+            locked_by = old_page.locked_by
+            locked_at = old_page.locked_at
+            locale = old_page.locale
+            seo_title = old_page.seo_title
+            search_description = old_page.search_description
+            show_in_menus = old_page.show_in_menus
```
[the remark this answers is not in the document]
Comment (engineervix) [reply tbx/propositions/management/commands/migrate_subservicepages.py]: Thanks, updated in b90a0852.

File: tbx/propositions/management/commands/migrate_subservicepages.py [hunk 5]
```diff
@@ -0,0 +1,1005 @@
...
+            if locked:
+                new_page.locked_by = locked_by
+                new_page.locked_at = locked_at
+
+            parent.add_child(instance=new_page)
+            new_page.save()
+
+            # Let's now deal with the content StreamField
+            self.show_status(f"Copying existing blocks from {old_page} ...")
+            copy_existing_blocks(old_page, new_page)
```
[the remark this answers is not in the document]
Comment (engineervix) [reply tbx/propositions/management/commands/migrate_subservicepages.py]: Thanks, updated in b016eb68, fd6c401f & e7210058

File: tbx/propositions/management/commands/migrate_subservicepages.py [hunk 6]
```diff
@@ -0,0 +1,1005 @@
...
+            construct_testimonials_block(old_page, new_page)
+
+            self.show_status(f"Constructing work block from {old_page} ...")
+            construct_work_block(old_page, new_page)
+
+            self.show_status(f"Constructing thinking block from {old_page} ...")
+            construct_thinking_block(old_page, new_page)
+
+            # Create a migration record
+            MigrationRecord.objects.create(
```
[the remark this answers is not in the document]
Comment (engineervix) [reply tbx/propositions/management/commands/migrate_subservicepages.py:969]: Good point, updated as part of previous comment, and also in 75ff893c

File: tbx/propositions/management/commands/undo_subservicepage_migration.py [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+                # Delete the migration record
+                self.show_status(f"Now deleting {migration_record}...")
+                migration_record.delete()
+                num_migration_records_deleted += 1
+
+        # Update links
+        self.show_status("Now reverting changes to links in richtext fields ...")
+        update_rich_text_links()
+        self.show_status("Now reverting changes to links in streamfields ...")
+        pages_to_manually_check = update_page_links()
```
[the remark this answers is not in the document]
Comment (engineervix) [reply tbx/propositions/management/commands/undo_subservicepage_migration.py]: Thanks! Gotta keep things DRY! Updated in ee4b1bb2 and febcc2b0

**Comment by engineervix** [comment]
> Hi Victor,
> 
> Wow, what a lot of code here. Well done for tackling it.
> 
> When testing locally with staging data, I encountered an error to do with the old testimonial streamfield (`AttributeError: 'StructValue' object has no attribute 'quote'`). I think we need to be more pessimistic about the existing streamfield data. We may need more exception handling added.
> 
> When the error was encountered, the command just failed halfway. I reverted it, but the revert didn't work fully as the page with the error was never added to the migration table.
> 
> Hope it won't be too much work to make these changes, but as this concerns database changes, we need to be more cautious.

Thanks @victoriachan for taking the time to look into this, and thanks for the feedback. I've made some changes and tested again. If you're happy with the changes, I can backup staging DB then run this on staging for QA and verification.

I don't have access to the production DB, it would be great to pull the production DB and run the script against it so we see how it turns out.

File: tbx/propositions/management/commands/migrate_subservicepages.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (victoriachan) [reply tbx/propositions/management/commands/migrate_subservicepages.py]: Ah yes, makes sense. Thanks for clarifying!

File: tbx/propositions/management/commands/migrate_subservicepages.py [hunk 8]
```diff
@@ -0,0 +1,1005 @@
...
+                new_data[key] = new_value
+            else:
+                new_data[key] = update_stream_data(value)
+        return new_data
+
+    else:
+        return data
+
+
+def update_page_links():
```
[the remark this answers is not in the document]
Comment (victoriachan) [reply tbx/propositions/management/commands/migrate_subservicepages.py]: Yes, that's fair enough. That feature is rather new and has not been explored much yet. The documentation is not much either. Thanks for explaining!

File: tbx/propositions/management/commands/migrate_subservicepages.py [hunk 9]
```diff
@@ -0,0 +1,1232 @@
...
+                        del updated_data
+                    if "revision" in locals():
+                        del revision
+                    if "field" in locals():
+                        del field
+
+    return report_data
+
+
+def update_rich_text_links(source, target, reverse=False):
```
Comment (victoriachan) [inline tbx/propositions/management/commands/migrate_subservicepages.py:740]: See `update_stream_data`, the same applies here.

File: tbx/propositions/management/commands/migrate_subservicepages.py [hunk 10]
```diff
@@ -0,0 +1,1232 @@
...
+            merge_blocks(target, "testimonials", "testimonials", data["client_logos"])
+            merge_blocks(target, "testimonials", "testimonials", data["testimonials"])
+
+
+# -----------------------------------------------------------------------------
+# 3. Update links
+# -----------------------------------------------------------------------------
+
+
+def replace_link(match, ids, reverse=False):
```
Comment (victoriachan) [inline tbx/propositions/management/commands/migrate_subservicepages.py:507]: Can we pass in source and target instead of using reverse, to be consistent with the other functions? Also then we can use the `target` variable for `target_page_instance = get_migration_page(related_page_id=page_id, page_model=target)` as suggested by my comment below.

**Review by victoriachan — CHANGES_REQUESTED** [review]
Hi Victor,

Thanks for making the changes. I've got just one more request to improve the way the new pk is derived, so instead of matching by title and slug, we try to match from the migration record. Unless you know of a good reason why that won't work.

Also I tested locally with the prod data, it is hard to tell if it worked as it is hard to check all the references, and won't be immediately obvious if something is missed. I'm wondering if we need tests after all. 

The undo had some errors, one to do with segmentation fault, and another to do with 'DeferringRelatedManager'. There are 31 of these errors. When checking the pages manually, I can see the original reverted SubServicePage pages no longer have their previous references, even after I ran `rebuild_references_index`. 

See the truncated report below for example of errors:

```
29 migration records have been deleted successfully.
29 SubPropositionPages have been deleted successfully.
29 SubServicePages have been successfully reverted to their previous state
The following pages require manual checking of links to avoid 404s:
1.  An Extra 3 Million Organic Clicks a Month for the NHS Website
------------------------------------------------------------
    Page id: 1319
    Streamfield name: Unknown
    Error: Page excluded from migration due to segmentation fault

...

3.  Innovation 
------------------------------------------------------------
    Page id: 1881
    Streamfield name: featured_blog_posts
    Error: 'DeferringRelatedManager' object has no attribute 'get_prep_value'
```
